xblock/core.py

```diff
@@ -121,23 +159,530 @@ def open_local_resource(cls, uri):
+
+    def _clear_dirty_fields(self):
+        """
+        Remove all dirty fields from an XBlock.
+        """
+        self._dirty_fields.clear()
+
+    def _reset_dirty_field(self, field):
+        """
+        Resets dirty field value with the value from the field data cache.
+        """
+        if field in self._dirty_fields:
+            self._dirty_fields[field] = copy.deepcopy(
+                self._field_data_cache[field.name]
+            )
+
+    def add_xml_to_node(self, node):
+        """
+        For exporting, set data on `node` from ourselves.
+        """
+        # pylint: disable=E1101
+        # Set node.tag based on our class name.
+        node.tag = self.xml_element_name()
+        node.set('xblock-family', self.entry_point)
+
+        # Set node attributes based on our fields.
+        for field_name, field in list(self.fields.items()):
+            if field_name in ('children', 'parent', 'content'):
+                continue
+            if field.is_set_on(self) or field.force_export:
+                self._add_field(node, field_name, field)
+
+        # A content field becomes text content.
+        text = self.xml_text_content()
+        if text is not None:
+            node.text = text
+
+    def xml_element_name(self):
+        """
+        What XML element name should be used for this block?
+        """
+        return self.scope_ids.block_type
+
+    def xml_text_content(self):
+        """
+        What is the text content for this block's XML node?
+        """
+        if 'content' in self.fields and self.content:  # pylint: disable=unsupported-membership-test
+            return self.content
+        else:
+            return None
+
+    def _add_field(self, node, field_name, field):
+        """
+        Add xml representation of field to node.
+
+        Depending on settings, it either stores the value of field
+        as an xml attribute or creates a separate child node.
+        """
+        value = field.to_string(field.read_from(self))
+        text_value = "" if value is None else value
+
+        # Is the field type supposed to serialize the fact that the value is None to XML?
+        save_none_as_xml_attr = field.none_to_xml and value is None
+        field_attrs = {"none": "true"} if save_none_as_xml_attr else {}
+
+        if save_none_as_xml_attr or field.xml_node:
+            # Field will be output to XML as an separate element.
+            tag = etree.QName(XML_NAMESPACES["option"], field_name)
+            elem = etree.SubElement(node, tag, field_attrs)
+            if field.xml_node:
+                # Only set the value if forced via xml_node;
+                # in all other cases, the value is None.
+                # Avoids an unnecessary XML end tag.
+                elem.text = text_value
+        else:
+            # Field will be output to XML as an attribute on the node.
+            node.set(field_name, text_value)
+
+
+class XBlockMixin(Blocklike):
```

Comment:
If `XBlockMixin` is a subclass of `Blocklike` that overrides no other behavior, why not rename `Blocklike` to `XBlockMixin`? That would be more in line with our conventions, wouldn't it (where `XBlock` and `XBlockAside` both inherit from `XBlockMixin`)?

Reply:
So XBlockMixin is an unfortunate name. "Mixin" in the traditional sense means a class that isn't a meaningful base in itself, but is to be "mixed in" to other classes to add extra functionality. In that sense, the XBlockMixin isn't a mixin; rather, it's supposed to be a *base class for new custom XBlock mixins*. 

Contrast this with Blocklike, which is a meaningful base, one which I expect to be used all over xblock and edx-platform as a type annotation:

```python
def this_function_could_take_a_block_or_an_aisde(block: Blocklike):
    ...
```

...which I think reads better than using XBlockMixin as the annotation:

```python
def this_function_could_take_a_block_or_an_aisde(block: XBlockMixin):
    ...
```

I guess we could still reconcile the two classes by just aliasing `XBlockMixin = Blocklike`... what do you think @ormsbee ?

Reply:
Okay, you've convinced me. Based on your description, `XBlockBase` would also work for me, but I'm good with `Blocklike`. I do like the idea of the alias since I have no idea if someone might have built off of that in their fork/plugin.

Reply:
@ormsbee On second thought, I'm actually inclined to leave XBlockMixin as its own subclass of Blocklike. It can't put my finger on exactly why, but it seems like it'd be *less surprising* to have XBlockMixin be its own class rather than an alias. Like you say, XBlockMixin is part of the public API, so we can't just delete it.